app/models/petition.rb

```diff
@@ -737,7 +737,20 @@ def publish(time = Time.current)
 
     build_pe_number
 
-    update(state: OPEN_STATE, open_at: time, closed_at: closing_date(time))
+    transaction do
+      update!(state: OPEN_STATE, open_at: time, closed_at: closing_date(time))
+
+      unless collect_signatures?
+        if at_threshold_for_referral?
+          update!(referral_threshold_reached_at: time)
+        end
+
+        close!(time)
+        refer_or_reject!(time)
+      else
+        true
```

Comment:
Ah, forgot the return value - generally `unless ... else ... end` is considered an anti-pattern in Ruby so probably best to re-write like this:

```
if !collect_signatures?
  if at_threshold_for_referral?
    update!(referral_threshold_reached_at: time)
  end

  close!(time)
  refer_or_reject!(time)
else
  true
end
```